src/formula.jl

```diff
@@ -0,0 +1,119 @@
+using Tables
+using Statistics: mean
+using StatsModels:
+    @formula, hasintercept, modelcols, FormulaTerm, apply_schema, schema, coefnames
+
+"""
+    center_predictors(X::AbstractMatrix)
+
+Centers the columns of a matrix `X` of predictors.
+Returns a tuple with:
+1. `μ_X`: 1xK `Matrix` of `Float64`s of the means of the K columns in the original `X`
+matrix.
+2. `X_centered`: A `Matrix` of `Float64`s with the same dimensions as the original matrix
+`X` with the columns centered on mean μ=0.
+
+# Arguments
+- `X::AbstractMatrix`: a matrix of predictors where rows are observations and columns are
+variables.
+
+# Examples
+```jldoctest
+julia> X = rand(3, 2)
+3×2 Matrix{Float64}:
+ 0.291245   0.604541
+ 0.0960436  0.609567
+ 0.536451   0.0184553
+
+julia> μ_X, X_centered = center_predictors(X);
+
+julia> μ_X
+1×2 Matrix{Float64}:
+ 0.307913  0.410854
+
+julia> X_centered
+3×2 Matrix{Float64}:
+ -0.0166683   0.193686
+ -0.21187     0.198713
+  0.228538   -0.392399
+```
+"""
+function center_predictors(X::AbstractMatrix)
+    μ_X = mapslices(mean, X; dims=1)
+    X_centered = similar(X, Float64)
+    for (idx, val) in enumerate(eachcol(X))
+        X_centered[:, idx] = val .- μ_X[1, idx]
+    end
+    return μ_X, X_centered
+end
+
+"""
+    make_yX(formula::FormulaTerm, data)
+
+Constructs the response y vector and the matrix X of predictors.
+
+Note that the original `StatsModels.jl` implementation adds an intercept column filled
+with `1`s, which we do not.
+
+Returns a tuple with:
+1. `y`: A `Vector` of the response variable in the `formula` and present inside `data`.
+2. `X`: A `Matrix` of the predictors variables in the `formula` and present inside `data`.
+
+# Arguments
+- `formula`: a `FormulaTerm` created by `StatsModels.@formula` macro.
+- `data`: `data` can either be a whole table (a property-accessible collection of iterable
+columns or iterable collection of property-accessible rows, as defined by
+[Tables.jl](https://github.com/JuliaData/Tables.jl), which includes `DataFrame`s) or a
+single row (in the form of a `NamedTuple` of scalar values).
```

Comment:
```suggestion
- `data`:  a `data` object that satisfies the [Tables.jl](https://github.com/JuliaData/Tables.jl) interface such as a DataFrame.
```
Because of the `Tables.istable(data)` check in the first line.